core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -494,12 +505,15 @@ namespace kagome::blockchain {
 
     } else {
       // ... or repair tree by parent of root
-      auto hash_res = header_repo_->getHashByNumber(node->depth - 1);
-      BOOST_ASSERT_MSG(hash_res.has_value(),
-                       "Non genesis block must have parent");
-
-      primitives::BlockInfo block{node->depth - 1, hash_res.value()};
-      auto tree = std::make_shared<TreeNode>(block.hash, block.number, true);
+      auto _header = header_repo_->getBlockHeader(node->depth - 1);
+      BOOST_ASSERT_MSG(_header, "Non genesis block must have parent");
+      auto &header = _header.value();
+      primitives::BlockInfo block{
+          node->depth - 1,
+          hasher_->blake2b_256(scale::encode(header).value()),
```

Comment:
Hm. Seems it consumes more resources (getting header + hash calc vs. getting hash). 
Why did you decide to use this way?

Reply:
We need header for `isPrimary`